File: examples/python/math-demo/MathDemo.runar.py
from runar import (
    StatefulSmartContract, Bigint, public, assert_,
    safediv, percent_of, clamp, sign, pow_, sqrt, gcd, mul_div, log2,
)

class MathDemo(StatefulSmartContract):
    value: Bigint

    def __init__(self, value: Bigint):
        super().__init__(value)
        self.value = value

    @public
    def divide_by(self, divisor: Bigint):
        self.value = safediv(self.value, divisor)

    @public
    def withdraw_with_fee(self, amount: Bigint, fee_bps: Bigint):
        fee = percent_of(amount, fee_bps)
        total = amount + fee
        assert_(total <= self.value)
        self.value = self.value - total

    @public
    def clamp_value(self, lo: Bigint, hi: Bigint):
        self.value = clamp(self.value, lo, hi)

    @public
    def normalize(self):
        self.value = sign(self.value)

    @public
    def exponentiate(self, exp: Bigint):
        self.value = pow_(self.value, exp)

    @public
    def square_root(self):
        self.value = sqrt(self.value)

    @public
    def reduce_gcd(self, other: Bigint):
        self.value = gcd(self.value, other)

    @public
    def scale_by_ratio(self, numerator: Bigint, denominator: Bigint):
        self.value = mul_div(self.value, numerator, denominator)

    @public
    def compute_log2(self):
        self.value = log2(self.value)

File: examples/ruby/math-demo/MathDemo.runar.rb
require 'runar'

class MathDemo < Runar::StatefulSmartContract
  prop :value, Bigint

  def initialize(value)
    super(value)
    @value = value
  end

  runar_public divisor: Bigint
  def divide_by(divisor)
    @value = safediv(@value, divisor)
  end

  runar_public amount: Bigint, fee_bps: Bigint
  def withdraw_with_fee(amount, fee_bps)
    fee = percent_of(amount, fee_bps)
    total = amount + fee
    assert total <= @value
    @value = @value - total
  end

  runar_public lo: Bigint, hi: Bigint
  def clamp_value(lo, hi)
    @value = clamp(@value, lo, hi)
  end

  runar_public
  def normalize
    @value = sign(@value)
  end

  runar_public exp: Bigint
  def exponentiate(exp)
    @value = pow(@value, exp)
  end

  runar_public
  def square_root
    @value = sqrt(@value)
  end

  runar_public other: Bigint
  def reduce_gcd(other)
    @value = gcd(@value, other)
  end

  runar_public numerator: Bigint, denominator: Bigint
  def scale_by_ratio(numerator, denominator)
    @value = mul_div(@value, numerator, denominator)
  end

  runar_public
  def compute_log2
    @value = log2(@value)
  end
end
